File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head

        groupPrev = dummy 
        while True:            
            kth = self.getKth(groupPrev, k)
            
            if not kth:
                break
            
            groupStart = groupPrev.next
            groupNext = kth.next

            cur = groupStart
            prev = groupNext

            while cur != groupNext:
                tmp = cur.next
                cur.next = prev
                prev = cur
                cur = tmp

            
            # groupPrev -> { 1(groupPrev.next), 2, 3(kth) } -> groupNext
            # groupPrev -> { 3, 2, 1} -> groupNext
            groupPrev.next = kth
            groupPrev = groupStart


        return dummy.next

    def getKth(self, curr, k):
        while curr and k > 0:
            k -= 1
            curr = curr.next

        return curr
    
    # dummy - 1 - 2 - 3  start: 1 2 3 K: 3 2 1 

File: Data Structures & Algorithms/add-two-numbers/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode cur = dummy;

        int carry = 0;

        while (l1 != null || l2 != null) {
            int v1;
            int v2;
            if (l1 == null) {
                v1 = 0;
                v2 = l2.val;
                l2 = l2.next;

            } else if (l2 == null) {
                v1 = l1.val;
                v2 = 0;
                l1 = l1.next;
            } else {
                v1 = l1.val;
                v2 = l2.val;
                l1 = l1.next;
                l2 = l2.next;
            }

            int v = (v1 + v2 + carry) % 10;
            carry = (v1 + v2 + carry) / 10;

            cur.next = new ListNode(v);
            cur = cur.next;
            
        }

        if (carry != 0) {
            cur.next = new ListNode(carry);
        }
        return dummy.next;

    }
}
